pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,115 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
 }
 
 // ValidateJobStatusUpdate validates an update to a JobStatus and returns an ErrorList with any errors.
-func ValidateJobStatusUpdate(status, oldStatus batch.JobStatus) field.ErrorList {
+func ValidateJobStatusUpdate(job, oldJob *batch.Job) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, validateJobStatus(&status, field.NewPath("status"))...)
+	statusFld := field.NewPath("status")
+	allErrs = append(allErrs, validateJobStatus(&job.Status, statusFld)...)
+
+	// A strengthened validation of the Job status transitions is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the transitions done by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// For example, we verify that a Job in terminal state (Failed or Complete)
+	// does not flip to a non-terminal state.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
```

Comment:
don't check feature gates in validation. If we want to conditionally be more strict about fields, pass in a validation options struct with boolean options for each thing we want to be more strict on, and populate the boolean based on the current state of the object.

PodValidationOptions is a good example of this, can follow the pattern where we pass that into validation and set the options based on the existing object in GetValidationOptionsFromPodSpecAndMeta

Reply:
Pushed the change to use `JobStatusValidationOptions`, similarly as we do for PodValidationOptions. I think the code is much cleaner, as we split the responsibility.